File: program/program/convert-standard-date-format-to-julian-date-and-vice-versa/ConvertStandardDateFormatToJulianDateAndViceVersa.ts
function standardToJulian(date: string): string {
    const [year, month, day] = date.split('-').map(Number);
    const julianDay = getDayOfYear(year, month, day);
    return `${year}-${pad(julianDay, 3)}`;
}

function julianToStandard(julianDate: string): string {
    const [year, dayOfYear] = julianDate.split('-').map(Number);
    const date = getDateFromDayOfYear(year, dayOfYear);
    const month = pad(date.getMonth() + 1, 2);
    const day = pad(date.getDate(), 2);
    return `${year}-${month}-${day}`;
}

function getDayOfYear(year: number, month: number, day: number): number {
    const date = new Date(year, month - 1, day);
    const startOfYear = new Date(year, 0, 0);
    const diff = +date - +startOfYear;
    const oneDay = 1000 * 60 * 60 * 24;
    return Math.floor(diff / oneDay);
}

function getDateFromDayOfYear(year: number, dayOfYear: number): Date {
    const date = new Date(year, 0); // January 1st
    date.setDate(dayOfYear);
    return date;
}

function pad(num: number, size: number): string {
    let str = num.toString();
    while (str.length < size) {
        str = '0' + str;
    }
    return str;
}

// Example usage:
const standardDate = '2022-02-22';
const julianDate = '22-053';

console.log(`Standard to Julian: ${standardToJulian(standardDate)}`);
console.log(`Julian to Standard: ${julianToStandard(julianDate)}`);

File: program/program/convert-standard-date-format-to-julian-date-and-vice-versa/convert_standard_date_format_to_julian_date_and_vice_versa.r
# Function to convert standard date format to Julian date
standard_to_julian <- function(date) {
  parts <- unlist(strsplit(date, "-"))
  year <- as.numeric(parts[1])
  month <- as.numeric(parts[2])
  day <- as.numeric(parts[3])
  
  julian_day <- as.numeric(format(as.Date(date), "%j"))
  julian_year <- as.numeric(format(as.Date(date), "%y"))
  
  return(paste(julian_year, sprintf("%03d", julian_day), sep = "-"))
}

# Function to convert Julian date format to standard date format
julian_to_standard <- function(julian_date) {
  parts <- unlist(strsplit(julian_date, "-"))
  julian_year <- as.numeric(parts[1])
  julian_day <- as.numeric(parts[2])
  
  standard_date <- as.Date(paste0("20", julian_year, "-01-01")) + julian_day - 1
  
  return(as.character(standard_date))
}

# Example usage:
standard_date <- "2022-02-22"
julian_date <- "22-053"

cat("Standard to Julian:", standard_to_julian(standard_date), "\n")
cat("Julian to Standard:", julian_to_standard(julian_date), "\n")
